File: src/algorithms/sliding-window.js
/**
 * SLIDING WINDOW ALGORITHM
 * 
 * Concept: Track exact timestamps of all requests within a moving time window.
 * Unlike fixed windows (which reset every minute), the window continuously
 * "slides" — it always represents the last N milliseconds.
 * 
 * Problem with Fixed Windows:
 * Limit = 100 req/min. At 12:00:59 a user sends 100 requests.
 * At 12:01:00 the window resets — they can send 100 MORE immediately.
 * In 2 seconds: 200 requests. The fixed window is fooled.
 * 
 * Sliding Window Fix:
 * At any moment, the window covers [now - 60s, now].
 * Those 100 requests at 12:00:59 still count at 12:01:00.
 * The user must wait until 12:01:59 to make more requests.
 * 
 * Trade-off: Slightly higher Redis memory (stores each request timestamp),
 * but provides much more accurate rate limiting.
 */

const redisClient = require('../redis-client');
const { v4: uuidv4 } = require('uuid');

class SlidingWindowLimiter {
  /**
   * @param {object} options
   * @param {number} options.windowMs     - Window size in milliseconds
   * @param {number} options.maxRequests  - Max requests per window
   * @param {string} options.keyPrefix    - Redis key namespace
   */
  constructor(options = {}) {
    this.windowMs = options.windowMs || 60000; // Default: 1 minute
    this.maxRequests = options.maxRequests || 100;
    this.keyPrefix = options.keyPrefix || 'rl:sliding_window';
  }

  /**
   * Check if a request is allowed under the sliding window algorithm.
   * 
   * @param {string} identifier - Unique identifier (user ID, IP, API key)
   * @returns {object} { allowed, remaining, retryAfter, limit }
   */
  async isAllowed(identifier) {
    const key = `${this.keyPrefix}:${identifier}`;
    const now = Date.now();
    // Unique ID ensures no two requests collide in the sorted set
    // even if they arrive at the exact same millisecond
    const requestId = `${now}-${uuidv4()}`;

    const result = await redisClient.evalScript('slidingWindow', [key], [
      this.windowMs,
      this.maxRequests,
      now,
      requestId,
    ]);

    // Lua returns: [allowed, remaining, retry_after_ms, max_requests]
    const [allowed, remaining, retryAfter, limit] = result;

    return {
      allowed: allowed === 1,
      remaining: Math.max(0, remaining),
      retryAfter: retryAfter > 0 ? Math.ceil(retryAfter / 1000) : 0,
      limit: limit,
      windowMs: this.windowMs,
      algorithm: 'sliding_window',
      identifier,
    };
  }

  /**
   * Get the current request count in the window (for monitoring).
   */
  async getWindowCount(identifier) {
    const key = `${this.keyPrefix}:${identifier}`;
    const client = redisClient.getClient();
    const now = Date.now();
    const windowStart = now - this.windowMs;

    // Count entries in the sorted set within the current window
    const count = await client.zcount(key, windowStart, now);

    return {
      count,
      remaining: Math.max(0, this.maxRequests - count),
      windowMs: this.windowMs,
      maxRequests: this.maxRequests,
    };
  }

  /**
   * Reset the sliding window for a specific identifier.
   */
  async reset(identifier) {
    const key = `${this.keyPrefix}:${identifier}`;
    await redisClient.getClient().del(key);
    return { reset: true, identifier };
  }
}

module.exports = SlidingWindowLimiter;

File: src/scripts/sliding_window.lua
-- SLIDING WINDOW ALGORITHM (Redis Lua Script)
-- More accurate than fixed windows — eliminates boundary burst spikes
--
-- How it works:
-- Uses a Redis Sorted Set where each member = request timestamp.
-- On every request:
--   1. Remove timestamps older than the window (they've "slid" out)
--   2. Count remaining entries = current request count
--   3. If under limit → add new timestamp, allow request
--   4. If at/over limit → reject request
--
-- KEYS[1] = rate limit key (e.g., "rl:sliding_window:ip:1.2.3.4")
-- ARGV[1] = window duration in milliseconds (e.g., 60000 = 1 minute)
-- ARGV[2] = max requests allowed in window
-- ARGV[3] = current timestamp (milliseconds)
-- ARGV[4] = unique request ID (for sorted set member uniqueness)

local key = KEYS[1]
local window_ms = tonumber(ARGV[1])
local max_requests = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local request_id = ARGV[4]

-- The window starts at (now - window_ms)
local window_start = now - window_ms

-- Step 1: Remove all entries older than the sliding window
-- ZREMRANGEBYSCORE removes members with score < window_start
redis.call('ZREMRANGEBYSCORE', key, '-inf', window_start)

-- Step 2: Count how many requests exist within the current window
local current_count = redis.call('ZCARD', key)

-- Step 3: Decide to allow or deny
local allowed = 0
local remaining = 0

if current_count < max_requests then
  -- Add this request timestamp to the sorted set
  -- Score = timestamp (used for range queries)
  -- Member = unique ID (to avoid collisions at same millisecond)
  redis.call('ZADD', key, now, request_id)
  allowed = 1
  remaining = max_requests - current_count - 1
else
  allowed = 0
  remaining = 0
end

-- Set TTL so the key auto-expires if no activity
redis.call('PEXPIRE', key, window_ms)

-- Get oldest request time (to calculate retry-after)
local retry_after = 0
if allowed == 0 then
  local oldest = redis.call('ZRANGE', key, 0, 0, 'WITHSCORES')
  if oldest[2] then
    -- How long until the oldest request slides out of the window
    retry_after = math.ceil((tonumber(oldest[2]) + window_ms - now))
  end
end

-- Return: [allowed (0/1), remaining, retry_after_ms, max_requests]
return {allowed, remaining, retry_after, max_requests}
